staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -3345,6 +3367,63 @@ type PodSpec struct {
 	// +listType=map
 	// +listMapKey=name
 	SchedulingGates []PodSchedulingGate `json:"schedulingGates,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,38,opt,name=schedulingGates"`
+
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the Pod is allowed to start. The resources
+	// will be made available to those containers which consume them
+	// by name.
+	//
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate.
+	//
+	// This field is immutable.
+	//
+	// +patchMergeKey=name
+	// +patchStrategy=merge,retainKeys
+	// +listType=map
+	// +listMapKey=name
+	// +featureGate=DynamicResourceAllocation
+	ResourceClaims []PodResourceClaim `json:"resourceClaims,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,39,rep,name=resourceClaims"`
+}
+
+// PodResourceClaim references exactly one ResourceClaim through a ClaimSource
+// and adds a name to it that is used inside a Pod.
+type PodResourceClaim struct {
+	// A name under which this resource can be referenced by the containers.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Source determines where to find the ResourceClaim.
+	Source ClaimSource `json:"source,omitempty" protobuf:"bytes,2,name=source"`
+}
+
+// ClaimSource describes a reference to a ResourceClaim.
+//
+// Exactly one of these fields should be set.  Consumers of this type must
+// treat an empty object as if it has an unknown value.
+type ClaimSource struct {
+	// ResourceClaimName is the name of a ResourceClaim object in the same
+	// namespace as this pod.
```

Comment:
I've been told that +optional is legacy and not needed when the json tag has `omitempty`.

Reply:
I've added it to the core api (internal and v1) and re-generated all files. The only effect was in `staging/src/k8s.io/api/core/v1/generated.proto` where the `+optional` was copied into the field comment.

I think it's safe to leave it out and better because it's just redundant.

Reply:
Or is it meant to for end-users, not the tooling? But then plain English might be better.

Reply:
I'm still unsure where it makes a difference. But as it could be relevant after the release when the online documentation gets generated and for the sake of consistency I'll add it.

Reply:
Done.

I also double-checked my "omitempty"s: ResourceClass.DriverName incorrectly had it and ObjectMeta lacked it - fixed.